zerver/lib/actions.py

```diff
@@ -1284,6 +1285,93 @@ def do_delete_user(user_profile: UserProfile) -> None:
+
+    do_deactivate_user(user_profile, acting_user=None)
+
+    subscribed_huddle_recipient_ids = set(
+        Subscription.objects.filter(
+            user_profile=user_profile, recipient__type=Recipient.HUDDLE
+        ).values_list("recipient_id", flat=True)
+    )
+    user_id = user_profile.id
+    personal_recipient = user_profile.recipient
+    realm = user_profile.realm
+    date_joined = user_profile.date_joined
+
+    with transaction.atomic():
+        # The strategy is that before calling user_profile.delete(), we need to
+        # reassign Messages  sent by the user to a dummy user, so that they don't
+        # get affected by CASCADE. We cannot yet create a dummy user with .id
+        # matching that of the user_profile, so the general scheme is:
+        # 1. We create a *temporary* dummy for the initial re-assignment of messages.
+        # 2. We delete the UserProfile.
+        # 3. We create a replacement dummy user with its id matching what the UserProfile had.
+        # 4. This is the intended, final replacement UserProfile, so we re-assign
+        #    the messages from step (1) to it and delete the temporary dummy.
+        random_token = secrets.token_hex(16)
+        temp_replacement_user = create_user(
+            email=f"temp_deleteduser{random_token}@{get_fake_email_domain(realm)}",
+            password=None,
+            realm=realm,
+            full_name=f"Deleted User {random_token}",
+            active=False,
+            is_mirror_dummy=True,
+            force_date_joined=date_joined,
+            create_personal_recipient=False,
+        )
+        Message.objects.filter(sender=user_profile).update(sender=temp_replacement_user)
+        user_profile.delete()
+
+        replacement_user = create_user(
+            force_id=user_id,
+            email=f"deleteduser{user_id}@{get_fake_email_domain(realm)}",
+            password=None,
+            realm=realm,
+            full_name=f"Deleted User {user_id}",
+            active=False,
+            is_mirror_dummy=True,
+            force_date_joined=date_joined,
+            create_personal_recipient=False,
+        )
+        # We don't delete the personal recipient to preserve  personal messages!
+        # Now, the personal recipient belong to replacement_user, because
+        # personal_recipient.type_id is equal to replacement_user.id.
+        replacement_user.recipient = personal_recipient
+        replacement_user.save(update_fields=["recipient"])
+
+        Message.objects.filter(sender=temp_replacement_user).update(sender=replacement_user)
+        temp_replacement_user.delete()
+
+        subs_to_recreate = [
+            Subscription(
+                user_profile=replacement_user,
+                recipient=recipient,
+                is_user_active=replacement_user.is_active,
+            )
+            for recipient in Recipient.objects.filter(id__in=subscribed_huddle_recipient_ids)
+        ]
```

Comment:
This is copy-pasted from `do_delete_user`. Not sure if there's a nicer way to refactor things to avoid code duplication - since these functions are so **kind of** similar with subtle differences in the ordering of steps etc....